src/main/rx/srxl2.c

```diff
@@ -0,0 +1,634 @@
+    response.crc_high = ((uint8_t *) &crc)[1];
+    response.crc_low = ((uint8_t *) &crc)[0];
+
+    srxl2RxWriteData(&response, sizeof(response));
+
+    return true;
+}
+
+void srxl2ProcessChannelData(const srxl2ChannelDataHeader* channel_data, rxRuntimeConfig_t *rxRuntimeConfig) {
+    if (channel_data->rssi >= 0) {
+        const int rssi_percent = channel_data->rssi;
+        setRssi(scaleRange(rssi_percent, 0, 100, 0, RSSI_MAX_VALUE), RSSI_SOURCE_RX_PROTOCOL);
+    } else {
+        const int rssi_dbm = channel_data->rssi;
+        (void) rssi_dbm;
+        // @todo come up with a scheme to scale power values to 0-1023 range
+        setRssi(RSSI_MAX_VALUE / 2, RSSI_SOURCE_RX_PROTOCOL);
+    }
+
+    if (channel_data->rssi == 0) {
+        global_result = RX_FRAME_FAILSAFE;
+    } else {
+        global_result = RX_FRAME_COMPLETE;
+    }
+
+    const uint16_t *frame_channels = (const uint16_t *) (channel_data + 1);
+
+    if (channel_data->channel_mask.u8.channels_0_7) {
+        for (size_t i = 0; i < 8; ++i) {
+            if (channel_data->channel_mask.u8.channels_0_7 >> i & 1) {
+                rxRuntimeConfig->channelData[i] = *frame_channels++;
+            }
+        }
+    }
+
+    if (channel_data->channel_mask.u8.channels_8_15) {
+        for (size_t i = 0; i < 8; ++i) {
+            if (channel_data->channel_mask.u8.channels_8_15 >> i & 1) {
+                rxRuntimeConfig->channelData[8 + i] = *frame_channels++;
+            }
+        }
+    }
+
+    if (channel_data->channel_mask.u8.channels_16_23) {
+        for (size_t i = 0; i < 8; ++i) {
+            if (channel_data->channel_mask.u8.channels_16_23 >> i & 1) {
+                rxRuntimeConfig->channelData[16 + i] = *frame_channels++;
+            }
+        }
+    }
+
+    if (channel_data->channel_mask.u8.channels_24_31) {
+        for (size_t i = 0; i < 8; ++i) {
+            if (channel_data->channel_mask.u8.channels_24_31 >> i & 1) {
+                rxRuntimeConfig->channelData[24 + i] = *frame_channels++;
+            }
+        }
+    }
+
+    // DEBUG("channel data: %d %d %x\r\n",
+    //     channel_data_header->rssi, channel_data_header->frame_losses, channel_data_header->channel_mask.u32);
+}
+
+bool srxl2ProcessControlData(const srxl2ControlDataSubHeader* control_data, rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    const uint8_t own_id = FlightController << 4 | unit_id;
+    if (control_data->reply_id == own_id) {
+        telemetry_requested = true;
+//        DEBUG("command: %x reply_id: %x own_id: %x\r\n", control_data->command, control_data->reply_id, own_id);
+    }
+
+    switch (control_data->command) {
+        case ChannelData: {
+            srxl2ProcessChannelData((const srxl2ChannelDataHeader *) (control_data + 1), rxRuntimeConfig);
+        } break;
+
+        case FailsafeChannelData: {
+            srxl2ProcessChannelData((const srxl2ChannelDataHeader *) (control_data + 1), rxRuntimeConfig);
+            setRssiDirect(0, RSSI_SOURCE_RX_PROTOCOL);
```

Comment:
So this only happens for the failsafe case. The receiver does send a rssi value of 0 with failsafe data, but it is being set by the setRssi function which goes through filtering from what I can tell. Does it make sense to make the failsafe rssi value go through the filter instead of just directly setting the value?